### DISCUSSION TITLE
Branches differ after merge

### DISCUSSION THREAD

**Discussion by jrett** [community/community#54586] [Pull Requests] [answered]
### Select Topic Area

Question

### Body

I've been a user of git for many years, but not any kind of expert. I've been on GitHub for a few years, as a basic user. I've had experience with bitbucket, gitlab, azure devops, and probably some others that I can't remember.

I'm moving my organizations code base from ancient CVS to git, and GitHub. We've agreed on a workflow where we have two permanent branches, dev and release. (dev is the default)

I have it mostly up and running, but I keep having issues between the two permanent branches as when I pr and merge dev into release, it seems fine, but then when I look at the branches in GitHub, one of them is always ahead of the other. When I look at the diff, it shows the dev branch is needing to be updated. It shows the changes on the dev branch as not existing, as though it's comparing to an older commit.

After merging this back into dev, it does this again with the release branch. It shows one of them is behind the other, and it offers up the merge request button. This time when I look at the diff that it shows me, it clearly states that there are 0 differences, yet it's still wanting me to make a pull request to get them in sync.

I can go back and forth merging from one to the other repeatedly, it's never satisfied. 

What the fasfdajsk is going on?

**Comment by airtower-luna** [comment, marked as answer]
When you merge via the GitHub web interface, you get a *merge commit* on top of the target branch. So there's a new commit the other branch doesn't have. Based on your description that's likely all.

To avoid the merge commit you need to do a fast-forward merge. That means to just add new commits from the branch being merged to the target branch, and is only possible if the target branch doesn't have new changes of its own (the HEAD commit of the target branch is an *ancestor* of the branch you're merging from). After a fast-forward merge both branches will be at the same commit.

You can't do a fast-forward merge (ff-merge for short) over the web interface, you need to do it locally with `git merge` and then push. You can still create a PR if you need or want to, GitHub will detect when you push the ff-merge and mark it as merged.

See also: [Git Branching - Basic Branching and Merging](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging)

**Reply by jrett** [reply]
I guess that also explains why it keeps going back and forth endlessly between the two branches. That makes sense.

Anyone know if/when github plans to add a fast-forward merge to their UI? This is only an issue between the main and dev branches as they're both permanent branches, and how we use them we don't expect to do these merges very often. I'm concerned that I'm going to forget this conversation and forget to do it from the command line next time. It would be great if they add this.

In any case, thank you for your explanation.

**Reply by airtower-luna** [reply]
I completely agree about adding ff-merges to the PR UI, that would be very useful.

**Reply by ezleyton** [reply]
I've been using github almost daily for 12 or even more years, and never realized this happens because I barely use the branches page, now I know and im mad about this 😆
